### PR TITLE
[PF-163] Creates static custom route when network monitoring enabled

### PR DESCRIPTION
**Pull request by yonghaoy**
https://github.com/broadinstitute/gcp-dm-templates/blob/eeee90fa4619b273d07206a867b01914cdeb0a30/templates/network.py#L79

### CODE REVIEW

File: src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  private final GcpProjectConfig gcpProjectConfig;
+
+  public CreateRouteStep(CloudComputeCow computeCow, GcpProjectConfig gcpProjectConfig) {
+    this.computeCow = computeCow;
+    this.gcpProjectConfig = gcpProjectConfig;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext) throws RetryException {
+    if (!checkEnableNetworkMonitoring(gcpProjectConfig)) {
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java:49]: missed this last time around, but `check` to me implies a little more than a boolean evaluation. How about `isNetworkMonitoringEnabled`?

File: src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      return StepResult.getStepResultSuccess();
+    }
+    String projectId = flightContext.getWorkingMap().get(GOOGLE_PROJECT_ID, String.class);
+    try {
+      Network network =
+          getResource(() -> computeCow.networks().get(projectId, NETWORK_NAME).execute(), 404)
+              .get();
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java:56]: this is doing an unqualified .get() on an optional. worth an explanation on why that's ok and the optional can't be empty.

File: src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext) throws RetryException {
+    if (!checkEnableNetworkMonitoring(gcpProjectConfig)) {
+      return StepResult.getStepResultSuccess();
+    }
+    String projectId = flightContext.getWorkingMap().get(GOOGLE_PROJECT_ID, String.class);
+    try {
+      Network network =
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/CreateRouteStep.java:54]: nit, move this below the if routes exist check to be closer to its use in creating the route

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java [hunk 4]
```diff
@@ -119,6 +121,7 @@ public void testCreateGoogleProject_enableNetworkMonitoring() throws Exception {
     Project project = assertProjectExists(pool);
     assertNetworkExists(project);
     assertSubnetsExist(project, NetworkMonitoring.ENABLED);
+    assertRouteExist(project);
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:124]: probably also worth asserting the route doesn't exist in the base case.

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java [hunk 5]
```diff
@@ -254,6 +257,22 @@ protected void addStep(Step step) {
     }
   }
 
+  /** A {@link Flight} that has multiple route creation steps. */
+  public static class MultiRouteStepFlight extends GoogleProjectCreationFlight {
+    public MultiRouteStepFlight(FlightMap inputParameters, Object applicationContext) {
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:262]: here's another one. what was wrong with my Class<? extends Step> suggestion?

**Review by wchamber — APPROVED** [review]
couple nits

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java (hunk 5 above)
Comment (yonghaoy) [reply src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:262]: Tried few ways but seems it is not working. 
I remeber the last thing I tried is 

```
public static class MultiInstanceStepFlight extends GoogleProjectCreationFlight {
  private final Class<? extends Step> multiStepClazz;
  public MultiInstanceStepFlight(Class<? extends Step> multiStepClazz) {...}
...
  if (multiStepClazz.isInstance(step))
...
```

But like I said  https://github.com/DataBiosphere/terra-rbs/pull/34#discussion_r512176996, Flight is not instinciated in RBS, so I don't know where to passed parameters then... Unless one proposal I asked in that comment threads is to have some static setter like: 
```

  public static class MultiInstanceStepFlight extends GoogleProjectCreationFlight {
    private static Class<? extends Step> multiStepClazz;
    public MultiInstanceStepFlight(FlightMap inputParameters, Object applicationContext) {
      super(inputParameters, applicationContext);
    }
    
    public static void setStepClass(Class<? extends Step>  clazz) {
      multiStepClazz = clazz;
    }
    @Override
    protected void addStep(Step step) {
      super.addStep(step);
      if (multiStepClazz.isInstance(step)) {
        super.addStep(step);
      }
    }
  }
```

DO you think that is okay or is there any other better solutions?

Comment (yonghaoy) [reply src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:264]: @wchamber

Comment (wchamber) [reply src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:264]: oh, got it. It's kind of awkward Stairway really wants to use constructors here. Registering creation functions would be nicer.
We could put the Class in the `FlightMap inputParameters`.

**Review by wchamber — APPROVED** [review]

**Pull request merged**